### ISSUE TITLE
FS#3373 - IPV6 flow offload broken

### ISSUE THREAD

**Issue by openwrt-bot** [openwrt/openwrt#8239]
*graphine:*

Enabling this option makes the ipv6 connection unstable.

Example: connecting to #openwrt-devel from hexchat (Ubuntu) makes you constantly disconnect. 
Disabling flow offload makes the connection stable.

**Comment by Shine-** [comment]
As mentioned above, this issue is fixed for version 21.02 (kernel 5.4) by PR #4849.

**Comment by SebDanielsson** [comment]
PR still open. Is it safe to activate software offloading or do we need to wait for merge?

**Comment by Headcrabed** [comment]
So why the patch haven't been merged? The patch have been there for a long time.

**Comment by Leo-PL** [comment]
I think it needs more testing, and not everyone is capable of testing that - the problem manifests on native IPv6 connections.

**Comment by swg0101** [comment]
On master commit af434e0da2485bd7a82895b5bb63b1182154b98e - getting about 15% packet loss when offload is turned on on my R7800 on kernel 5.10.107, so definitely not just a 5.4 problem.

**Comment by bennyzhou88** [comment]
Since IPv6 doesn't involve NAT, turn the "flow-offloading" on will not improve the IPv6 performance. Only IPv4 NAT needs "flow-offloading". 

My solution is to turn on the HW-NAT for IPv4 only

**Comment by swg0101** [comment]
> Since IPv6 doesn't involve NAT, turn the "flow-offloading" on will not improve the IPv6 performance. Only IPv4 NAT needs "flow-offloading".
> 
> My solution is to turn on the HW-NAT for IPv4 only

I don't believe that's true at all. Enabling software offloading creates a flowable in which more or less the flow configuration is saved. While NAT is part of that decision, stuff like routing decisions and interfaces are also cached. I think this page explains quite clearly how it works:
https://www.kernel.org/doc/html/v5.8/networking/nf_flowtable.html

**Comment by bennyzhou88** [comment]
> > Since IPv6 doesn't involve NAT, turn the "flow-offloading" on will not improve the IPv6 performance. Only IPv4 NAT needs "flow-offloading".
> > My solution is to turn on the HW-NAT for IPv4 only
> 
> I don't believe that's true at all. Enabling software offloading creates a flowable in which more or less the flow configuration is saved. While NAT is part of that decision, stuff like routing decisions and interfaces are also cached. I think this page explains quite clearly how it works: https://www.kernel.org/doc/html/v5.8/networking/nf_flowtable.html

I agree with you. But IPv6 native routing doesn't get as much benefit as IPv4 NAT with "flow-offloading".

Since this FS#3373 issue is unsolved by now, the best I can do is to leave the IPv6 flow-offload off and only enable it for IPv4.

**Comment by Shine-** [comment]
bennyzhou88, any chance you could post numbers to prove your claim that flow offloading doesn't improve IPv6 performance for you?

If you look at my numbers [here](https://github.com/openwrt/openwrt/pull/4849#issuecomment-1079950052), I do see a HUGE difference. You may disagree, but 600% increase in throughput is definitely a huge difference in my opinion.

As mentioned earlier, FS#3373 is in fact fixed with PR #4849, at least for 21.02 (kernel 5.4). Unfortunately, it's not merged, so you have to patch and build on your own.

**Comment by MrObvious** [comment]
Does anyone have a build for Archer C7 v2 with this fix?

**Comment by woshilaiba** [comment]
is anyone knows if  this fixed in 22.03？

**Comment by Shine-** [comment]
Merged just yesterday into 22.03 branch.
21.02 is still unfixed (PR existing but not merged yet).

**Comment by Neverends4** [comment]
> Does anyone have a build for Archer C7 v2 with this fix?

download 22.03 rc4 you will have this patch automatically.

**Comment by dR3b** [comment]
I can confirm this on:
```bash
Model | MikroTik RouterBOARD 760iGS
Architecture | MediaTek MT7621 ver:1 eco:3
Target Platform | ramips/mt7621
OpenWrt 21.02.3 r16554-1d4dea6d4f / LuCI openwrt-21.02 branch git-22.213.35964-87836ca
5.4.188
```

**Comment by cbeyls** [comment]
My TP-Link Archer C7 V5 shows abysmal performance on 22.03 with flow offloading enabled (~100 Mbps vs ~450 Mbps on 19.07). I have yet to identify the cause.
Would it be possible to have the patch merged in a future 21.02 stable release ?

**Comment by dR3b** [comment]
@cbeyls OpenWrt 22.03 is stable now!

**Comment by cbeyls** [comment]
@dR3b OpenWRT 22.03 is stable but has low performance on my router compared to previous releases so I would rather install 21.02 for now

**Comment by Shine-** [comment]
The fix was merged into 21.02 with #9940 two months ago already.
Perhaps a dev can close this issue, so people will realize it's fixed for all affected versions by now?

**Comment by hauke** [comment]
This was fixed
OpenWrt master: efff48529b5cdb4046b923bebee483c7c68755a7
OpenWrt 22.03: 972160a0f7526fad1be9529879f16791a33355f9
OpenWrt 21.02: 6d891adf390dd232c63fa77a8e71750495c04159
